File: source/go/map_reduce.go
package main

import (
	"fmt"
	"sync"
)

func mapWorker(chunk []int, resultCh chan<- []int, wg *sync.WaitGroup) {
	defer wg.Done()
	var results []int
	for _, v := range chunk {
		results = append(results, v*v)
	}
	resultCh <- results
}

func reduceWorker(values []int) int {
	sum := 0
	for _, v := range values {
		sum += v
	}
	return sum
}

func main() {
	data := []int{1, 2, 3, 4, 5, 6, 7, 8}
	chunkSize := 2

	var chunks [][]int
	for i := 0; i < len(data); i += chunkSize {
		end := i + chunkSize
		if end > len(data) {
			end = len(data)
		}
		chunks = append(chunks, data[i:end])
	}

	resultCh := make(chan []int, len(chunks))
	var wg sync.WaitGroup

	for _, chunk := range chunks {
		wg.Add(1)
		go mapWorker(chunk, resultCh, &wg)
	}

	wg.Wait()
	close(resultCh)

	var allSquares []int
	for result := range resultCh {
		allSquares = append(allSquares, result...)
	}

	total := reduceWorker(allSquares)
	fmt.Printf("sum of squares: %d\n", total)
}

File: source/python/map_reduce.py
import threading

data = [1, 2, 3, 4, 5, 6, 7, 8]

def map_worker(chunk, result_holder):
    result_holder.extend(x * x for x in chunk)

def reduce_worker(values, result_holder):
    result_holder.append(sum(values))

chunk_size = 2
chunks = [data[i:i + chunk_size] for i in range(0, len(data), chunk_size)]

map_results = [[] for _ in chunks]
map_threads = []
for i, chunk in enumerate(chunks):
    t = threading.Thread(target=map_worker, args=(chunk, map_results[i]))
    t.start()
    map_threads.append(t)

for t in map_threads:
    t.join()

all_squares = []
for r in map_results:
    all_squares.extend(r)

reduce_result = []
reduce_thread = threading.Thread(target=reduce_worker, args=(all_squares, reduce_result))
reduce_thread.start()
reduce_thread.join()

print(f"sum of squares: {reduce_result[0]}")
